torch/nn/functional.py

```diff
@@ -571,6 +571,37 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
     return f(input, weight, bias)
 
 
+def relation(input, g, f=None, embedding=None, max_pairwise=None):
+    # Batch size, number of objects, feature size
+    b, o, c = input.size()
+    # Construct pairwise indices
+    i = Variable(torch.arange(0, o).type_as(input.data).long().repeat(o))
+    j = Variable(torch.arange(0, o).type_as(input.data).long().repeat(o, 1).t().contiguous().view(-1))
+    # Create pairwise matrix
+    pairs = torch.cat((torch.index_select(input, 1, i), torch.index_select(input, 1, j)), 2)
+    # Append embedding if provided
+    if embedding is not None:
+        pairs = torch.cat((pairs, embedding.unsqueeze(1).expand(b, o ** 2, embedding.size(1))), 2)
+    # Calculate new feature size
+    c = pairs.size(2)
+    # Pack into batches
+    pairs = pairs.view(b * o ** 2, c)
+    # Pass through g
+    if max_pairwise is None:
+        output = g(pairs)
+    else:
+        outputs = []
+        for batch in range(0, b * o ** 2, max_pairwise):
+            outputs.append(g(pairs[batch:batch + max_pairwise]))
```

Comment:
Agreed that the graph still has to be built (in the non-volatile case) so we're not saving memory, but I *think* it's a computation thing? Pinging @imisra to see if what I've done is useful - otherwise we remove this optional argument to simplify things even further and think about optimising it later.